### PR TITLE
BTM-827 Issue with ingesting invoices with multiple lines #486

### PR DESCRIPTION
**Pull request by mark-cab**

### CODE REVIEW

**Review by MLMartinDG — APPROVED** [review]

File: src/handlers/store-standardised-invoices/handler.ts [hunk 1]
```diff
@@ -24,23 +26,52 @@ export const handler = async (event: SQSEvent): Promise<Response> => {
...
     throw new Error("Raw invoice bucket not set.");
 
+  for (const records of event.Records) {
+    if (!records || !records.body) {
+      throw new Error("Record contains no body");
+    }
+  }
+
+  let recordBodies: StandardisedLineItem[] = [];
+  for (const records of event.Records) {
```
Comment (MLMartinDG) [inline src/handlers/store-standardised-invoices/handler.ts:36]: Can we merge the above for loop into this one?

File: src/handlers/store-standardised-invoices/handler.ts [hunk 2]
```diff
@@ -24,23 +26,52 @@ export const handler = async (event: SQSEvent): Promise<Response> => {
...
+    const recordList: StandardisedLineItem[] = JSON.parse(records.body);
+    recordBodies = recordBodies.concat(recordList);
+  }
```
Comment (Si3rr4wow) [inline src/handlers/store-standardised-invoices/handler.ts:39]:
```suggestion
  const recordBodies = flatMap(record => JSON.parse(records.body))
```

might be a more javascripty way to do this

File: src/handlers/store-standardised-invoices/handler.ts (hunk 1 above)
Comment (mark-cab) [reply src/handlers/store-standardised-invoices/handler.ts:36]: would rather keep the validation seperate to be honest, will update second one though as it's not needed as much as it's not got lots of logs in it now

File: src/handlers/store-standardised-invoices/handler.ts [hunk 3]
```diff
@@ -24,23 +26,52 @@ export const handler = async (event: SQSEvent): Promise<Response> => {
...
+      logger.error(`Handler failure for ${JSON.stringify(body)}`, { error });
+      response.batchItemFailures.push({
+        itemIdentifier: JSON.stringify(body),
+      });
     }
   });
 
   await Promise.all(promises);
+
+  if (response.batchItemFailures.length === 0) {
```
Comment (Si3rr4wow) [inline src/handlers/store-standardised-invoices/handler.ts:60]: AWS::Serverless::Api

Comment (mark-cab) [reply src/handlers/store-standardised-invoices/handler.ts:57]: not sure I follow, if some have succeeded and some have failed we don't want to move the file to the success folder we would want to reprocess it

File: src/handlers/store-standardised-invoices/store-line-item.test.ts [hunk 4]
```diff
@@ -249,15 +238,15 @@ describe("Line item storer", () => {
   });
 
   test("Line item storer with record body object with no invoice-is-quarterly option", async () => {
-    delete givenRecordBodyObject.invoice_is_quarterly;
-    givenRecord = { body: JSON.stringify(givenRecordBodyObject) } as any;
+    // @ts-expect-error
+    delete givenRecord.invoice_is_quarterly;
+    givenRecord = { body: JSON.stringify(givenRecord) } as any;
```
[the remark this answers is not in the document]
Comment (mark-cab) [reply src/handlers/store-standardised-invoices/store-line-item.test.ts:243]: why the ts ignore? That field is not optional but delete requires optional so when I corrected the record object to use StandardisedLineItem type this line gave a ts error

File: src/handlers/store-standardised-invoices/handler.ts (hunk 3 above)
Comment (mark-cab) [reply src/handlers/store-standardised-invoices/handler.ts:57]: This  call is for moving the invoice file from the bucket to /successful. The files we create per line for further processing are saved as part of store-line-item.ts (putTextS3)

**Review by MLMartinDG — APPROVED** [review]

**Pull request merged**